File: Utilities/Matlab/scripts/USFS_Deep_Fuel_Beds.m

% 6/24/2022 Noelle Crump

% This code is set up to automate the production of a set of
% simulation Validation plots.  A list of parameters for each experiment is read
% in from a file. The units are not standard SI.

% Here are the columns:
% 1 run# -- a sequential number to track the simulation.
% 2 burn# -- the number of the actual deep fuel beds burn, if applicable
% 3 spacing -- between center points of fuel arrays in cm.
% 4 angle -- angle in degrees of inclined platform.
% 5 depth -- in inches.
% 6 moisture content -- in percent, dry weight basis, so mc = 5 = (wet-dry)/(dry).
% 9 temp -- ambient temp in degrees fahrenheit
% 10 rh -- relative humidity, in percent
% 11 burned -- 1 or zero; 1 means "burned", zero means "did not burn".
% 12 pct_burn -- percent of fuel burned.

close all
clear all

% Runs from Matlab folder (Validation plots implementation)
outdir = '../../../out/USFS_Deep_Fuel_Beds/';
expdir = '../../../exp/USFS_Deep_Fuel_Beds/';
plot_dir = '../../Manuals/FDS_Validation_Guide/SCRIPT_FIGURES/USFS_Deep_Fuel_Beds/';

EXP = importdata([expdir,'exp_params.csv'],',',1);

BURN_NO      = EXP.data(:,find(strcmp(EXP.colheaders,'BURN_NO')));
SPACING      = EXP.data(:,find(strcmp(EXP.colheaders,'SPACING'))).*2.5;
DEPTH        = EXP.data(:,find(strcmp(EXP.colheaders,'DEPTH'))).*2.5;
FMC          = EXP.data(:,find(strcmp(EXP.colheaders,'MOISTURE_CONTENT')));
SLOPE        = EXP.data(:,find(strcmp(EXP.colheaders,'SLOPE')));
BURN_PCT_EXP = EXP.data(:,find(strcmp(EXP.colheaders,'BURN_PERCENT')));

% read in data
for i = 1:length(BURN_NO)

    CHID = ['burn',num2str(BURN_NO(i)),'_',num2str(DEPTH(i)),'D_',num2str(SLOPE(i)),'S_',num2str(SPACING(i)),'L'];
    HRR  = importdata([outdir,[CHID,'_cat_hrr.csv']],',', 2);
    DEVC = importdata([outdir,[CHID,'_cat_devc.csv']],',', 2);

    % Determine FDS (out) Burn Type:
    hrr_col = HRR.data(:,(strcmp(HRR.colheaders,'HRR'))); % (kw)
    time_col = HRR.data(:,(strcmp(HRR.colheaders,'Time'))); % (s)
    fuel_load = DEVC.data(2,(strcmp(DEVC.colheaders,'"Fuel Load"'))); % (kg)
    dry_fuel_mass = (1/(1+FMC(i)*0.01))*fuel_load; % (kg)
    total_fuel_energy = dry_fuel_mass*17425.0; % (kJ) 17425 = heat of combustion

    ignition_energy = trapz([0,10,20],[500*0.3*1.6,500*0.3*1.6,0]); % (kJ) hrr ramp
    total_fds_energy = trapz (time_col,hrr_col); % (kJ)

    BURN_PCT_FDS(i) = (total_fds_energy-ignition_energy)/total_fuel_energy * 100.;
end

figure
plot_style

set(gca,'Units',Plot_Units)
set(gca,'Position',[Scat_Plot_X Scat_Plot_Y Scat_Plot_Width Scat_Plot_Height])

h = scatter(BURN_PCT_EXP,BURN_PCT_FDS,10,'filled');
box on

hold on
plot([0,100],[0,100],'k-')

for i=1:length(BURN_NO)
    % add run number to each marker
    text(BURN_PCT_EXP(i),BURN_PCT_FDS(i),['     ',num2str(BURN_NO(i))],'FontName',Font_Name,'FontSize',2);
end

xlim([0,100])
ylim([0,100])

xlabel('Fuel Energy Burned EXP (%)','Interpreter',Font_Interpreter,'FontSize',Scat_Label_Font_Size)
ylabel('Fuel Energy Burned FDS (%)','Interpreter',Font_Interpreter,'FontSize',Scat_Label_Font_Size)

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Scat_Label_Font_Size)

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Scat_Paper_Width Scat_Paper_Height]);
set(gcf,'Position',[0 0 Scat_Paper_Width Scat_Paper_Height]);
print(gcf,'-dpdf',[plot_dir,'USFS_Deep_Fuel_Beds_Scatterplot']);

return

File: Utilities/Python/scripts/USFS_Deep_Fuel_Beds.py

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.integrate import trapezoid

# include FDS plot styles, etc.
import fdsplotlib

# Get plot style parameters
plot_style = fdsplotlib.get_plot_style('fds')

# Paths
outdir = '../../../out/USFS_Deep_Fuel_Beds/'
expdir = '../../../exp/USFS_Deep_Fuel_Beds/'
plot_dir = '../../Manuals/FDS_Validation_Guide/SCRIPT_FIGURES/USFS_Deep_Fuel_Beds/'

git_file =  os.path.join(outdir, f'burn100_60D_15S_35L_cat_git.txt')
version_string = fdsplotlib.get_version_string(git_file)
    
# Read experimental data
EXP = pd.read_csv(os.path.join(expdir, 'exp_params.csv'), header=0)
BURN_NO = EXP['BURN_NO'].values
SPACING = EXP['SPACING'].values * 2.5
DEPTH = EXP['DEPTH'].values * 2.5
FMC = EXP['MOISTURE_CONTENT'].values
SLOPE = EXP['SLOPE'].values
BURN_PCT_EXP = EXP['BURN_PERCENT'].values
BURN_PCT_FDS = np.zeros(len(BURN_NO))
    
for i in range(len(BURN_NO)):
    CHID = f'burn{int(BURN_NO[i])}_{int(DEPTH[i])}D_{int(SLOPE[i])}S_{int(SPACING[i])}L'
    HRR = pd.read_csv(os.path.join(outdir, f'{CHID}_cat_hrr.csv'), header=1)
    HRR.columns = HRR.columns.str.strip()
    DEVC = pd.read_csv(os.path.join(outdir, f'{CHID}_cat_devc.csv'), header=1)
    DEVC.columns = DEVC.columns.str.strip()
    # Determine FDS (out) Burn Type:
    hrr_col = HRR['HRR'].values  # (kW)
    time_col = HRR['Time'].values  # (s)
    # Note: adjust column name if actual CSV has quotes around "Fuel Load"
    fuel_load = DEVC.iloc[1].get('Fuel Load', DEVC.iloc[1].get('"Fuel Load"', np.nan))
    dry_fuel_mass = (1 / (1 + FMC[i] * 0.01)) * fuel_load  # (kg)
    total_fuel_energy = dry_fuel_mass * 17425.0  # (kJ)
    ignition_energy = trapezoid([500 * 0.3 * 1.6, 500 * 0.3 * 1.6, 0], [0, 10, 20])  # (kJ)
    total_fds_energy = trapezoid(hrr_col, time_col)  # (kJ)
    BURN_PCT_FDS[i] = (total_fds_energy - ignition_energy) / total_fuel_energy * 100.0

fig = fdsplotlib.plot_to_fig([0,100],[0,100],
                             x_min=0, x_max=100, y_min=0, y_max=100,
                             figure_size=(plot_style['Scat_Paper_Width'],plot_style['Scat_Paper_Height']),
                             plot_size=(plot_style['Scat_Plot_Width'],plot_style['Scat_Plot_Height']),
                             plot_origin=(plot_style['Scat_Plot_X'],plot_style['Scat_Plot_Y']),
                             revision_label=version_string,
                             x_label='Measured Fuel Consumption (\%)',
                             y_label='Predicted Fuel Consumption (\%)'
                             )

fdsplotlib.plot_to_fig(BURN_PCT_EXP, BURN_PCT_FDS, figure_handle=fig, marker_style='ko')
    
for i in range(len(BURN_NO)):
    plt.text(BURN_PCT_EXP[i], BURN_PCT_FDS[i], f'{"  "}+{int(BURN_NO[i])}', fontname=plot_style['Font_Name'], fontsize=6)

plt.savefig(os.path.join(plot_dir, 'USFS_Deep_Fuel_Beds_Scatterplot.pdf'), format='pdf')


